clang/lib/CIR/Dialect/IR/CIRTypes.cpp

```diff
@@ -95,15 +95,28 @@ void BoolType::print(mlir::AsmPrinter &printer) const {}
 //===----------------------------------------------------------------------===//
 
 Type StructType::parse(mlir::AsmParser &parser) {
+  const auto loc = parser.getCurrentLocation();
   llvm::SmallVector<mlir::Type> members;
   mlir::StringAttr id;
   bool body = false;
   bool packed = false;
   mlir::cir::ASTRecordDeclAttr ast = nullptr;
+  RecordKind kind;
 
   if (parser.parseLess())
     return {};
 
+  // FIXME(cir): break struct kinds into different types.
```

Comment:
What do you mean?

Reply:
Instead of having a single type to represent records, I believe it would be better to have separate types for different source language declarations such as `cir.class`, `cir.union`, and `cir.struct`.

Separating classes and structs might not present any benefit, but unions behave rather differently than other record types. Handling each type separately would facilitate lowering by allowing us to handle each record type individually.

It could be argued that this would be detrimental to codegen since it usually handles these types as generic records, but this could easily be amended with an MLIR interface that could abstract these types through a `RecordLike` interface.

Reply:
Gotcha, can you then change to `TODO(cir): in the future we should probably separate types for different source language declarations such as cir.class, cir.union, and cir.struct`.